src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1794,73 +1831,186 @@ int TheoryStringsRewriter::componentContains(std::vector<Node>& n1,
     unsigned diff = n1.size() - n2.size();
     for (unsigned i = 0; i <= diff; i++)
     {
-      bool check = false;
-      if (n1[i] == n2[0])
-      {
-        check = true;
-      }
-      else if (n1[i].isConst() && n2[0].isConst())
-      {
-        // could be suffix
-        CVC4::String s = n1[i].getConst<String>();
-        CVC4::String t = n2[0].getConst<String>();
-        if (t.size() < s.size() && s.suffix(t.size()) == t)
-        {
-          check = true;
-        }
-      }
-      if (check)
+      Node n1rb_first;
+      Node n1re_first;
+      // first component of n2 must be a suffix
+      if (componentContainsBase(
+              n1[i], n2[0], n1rb_first, n1re_first, 1, computeRemainder))
       {
-        bool success = true;
+        Assert(n1re_first.isNull());
         for (unsigned j = 1; j < n2.size(); j++)
         {
-          if (n1[i + j] != n2[j])
+          // are we in the last component?
+          if (j + 1 == n2.size())
           {
-            if (j + 1 == n2.size() && n1[i + j].isConst() && n2[j].isConst())
+            Node n1rb_last;
+            Node n1re_last;
+            // last component of n2 must be a prefix
+            if (componentContainsBase(n1[i + j],
+                                      n2[j],
+                                      n1rb_last,
+                                      n1re_last,
+                                      -1,
+                                      computeRemainder))
             {
-              // could be prefix
-              CVC4::String s = n1[i + j].getConst<String>();
-              CVC4::String t = n2[j].getConst<String>();
-              if (t.size() < s.size() && s.prefix(t.size()) == t)
+              Assert(n1rb_first.isNull());
+              if (computeRemainder)
               {
-                if (computeRemainder)
+                if (remainderDir != -1)
                 {
-                  if (s.size() > t.size())
+                  if (!n1re_last.isNull())
                   {
-                    nr.push_back(
-                        NodeManager::currentNM()->mkConst(::CVC4::String(
-                            s.substr(t.size(), s.size() - t.size()))));
+                    ne.push_back(n1re_last);
+                  }
+                  ne.insert(ne.end(), n1.begin() + i + j + 1, n1.end());
+                  n1.erase(n1.begin() + i + j + 1, n1.end());
+                }
+                n1[i + j] = n2[j];
+                n1[i] = n2[0];
```

Comment:
Same issue as above.

Reply:
Also same argument as above :)

Reply:
The fix for the issue above looks good but I think that the same issue applies here.

Reply:
Nvm, after discussing this offline, I am convinced that this is ok :)